File: runtime/java/prompty/src/main/java/com/microsoft/prompty/References.java
package com.microsoft.prompty;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import org.yaml.snakeyaml.error.YAMLException;

/**
 * Resolution of {@code ${protocol:value}} references in loaded frontmatter.
 *
 * <p>Two protocols are recognised:
 *
 * <ul>
 *   <li>{@code ${env:VAR}} and {@code ${env:VAR:default}} — an environment variable, with an
 *       optional literal default used when the variable is unset. The default is everything after
 *       the first colon, so it may itself contain colons.
 *   <li>{@code ${file:path}} — the contents of a file resolved relative to the prompt's directory.
 *       {@code .json}, {@code .yaml} and {@code .yml} are parsed into structured values; anything
 *       else is inlined as raw text.
 * </ul>
 *
 * <p>Any other protocol is left untouched, so unrelated {@code ${...}} syntax in a template survives
 * loading.
 *
 * <h2>File sandboxing</h2>
 *
 * <p>A {@code ${file:...}} target must resolve inside the prompt's own directory or one of the
 * explicitly allowed roots from {@link LoadOptions}. This is enforced after canonicalisation, so
 * {@code ../} traversal and symlinks cannot escape.
 *
 * <h2>A deliberate quirk</h2>
 *
 * <p>{@link #resolveReferences} descends into nested maps and lists, but it only resolves strings
 * that are <em>map values</em> — a bare string sitting directly in a list is left alone. That
 * matches the Rust runtime exactly, and the runtimes are kept identical here rather than
 * independently "fixed", because the wire behaviour is observable.
 */
public final class References {

  private References() {}

  /**
   * Recursively resolve references in place, throwing on an unresolvable one.
   *
   * @param value the frontmatter tree; maps and lists are walked, other values are left alone
   * @param agentDir the prompt file's directory, always an allowed {@code ${file:}} root
   * @param allowedFileRoots additional directories {@code ${file:}} may read from
   */
  public static void resolveReferences(Object value, Path agentDir, List<Path> allowedFileRoots) {
    if (value instanceof Map<?, ?> rawMap) {
      @SuppressWarnings("unchecked")
      Map<String, Object> map = (Map<String, Object>) rawMap;
      for (String key : new ArrayList<>(map.keySet())) {
        Object item = map.get(key);
        if (item instanceof String s) {
          Optional<Object> resolved = tryResolveString(s, key, agentDir, allowedFileRoots, true);
          if (resolved.isPresent()) {
            map.put(key, resolved.get());
          }
        } else if (item != null) {
          resolveReferences(item, agentDir, allowedFileRoots);
        }
      }
    } else if (value instanceof List<?> rawList) {
      @SuppressWarnings("unchecked")
      List<Object> list = (List<Object>) rawList;
      for (int i = 0; i < list.size(); i++) {
        Object item = list.get(i);
        if (item instanceof String s) {
          // Spec §4.2 resolves every string value, not only the ones that happen to sit directly
          // under a key. A list of file references is a perfectly ordinary thing to write.
          Optional<Object> resolved =
              tryResolveString(s, "[" + i + "]", agentDir, allowedFileRoots, true);
          if (resolved.isPresent()) {
            list.set(i, resolved.get());
          }
        } else if (item != null) {
          resolveReferences(item, agentDir, allowedFileRoots);
        }
      }
    }
  }

  /**
   * Resolve a single string reference, swallowing failures.
   *
   * <p>Used as the model layer's {@code preProcess} hook, where the tree has normally already been
   * resolved by {@link #resolveReferences} and a second, throwing pass would be redundant.
   *
   * @return the resolved value, or empty if the string was not a resolvable reference
   */
  public static Optional<Object> resolveSingleRef(
      String value, Path agentDir, List<Path> allowedFileRoots) {
    return tryResolveString(value, "<preProcess>", agentDir, allowedFileRoots, false);
  }

  private static Optional<Object> tryResolveString(
      String value, String key, Path agentDir, List<Path> allowedFileRoots, boolean throwOnError) {
    if (!value.startsWith("${") || !value.endsWith("}")) {
      return Optional.empty();
    }

    String inner = value.substring(2, value.length() - 1);
    int colon = inner.indexOf(':');
    if (colon < 0) {
      return Optional.empty();
    }

    String protocol = inner.substring(0, colon).toLowerCase(Locale.ROOT);
    String rest = inner.substring(colon + 1);

    try {
      return switch (protocol) {
        case "env" -> resolveEnv(rest, key);
        case "file" -> resolveFile(rest, agentDir, allowedFileRoots, key);
        // Unknown protocol — leave the string as authored.
        default -> Optional.empty();
      };
    } catch (LoadException e) {
      if (throwOnError) {
        throw e;
      }
      return Optional.empty();
    }
  }

  private static Optional<Object> resolveEnv(String spec, String key) {
    int colon = spec.indexOf(':');
    String varName = colon < 0 ? spec : spec.substring(0, colon);
    String defaultValue = colon < 0 ? null : spec.substring(colon + 1);

    Optional<String> actual = Environment.lookup(varName);
    if (actual.isPresent()) {
      return Optional.of(actual.get());
    }
    if (defaultValue != null) {
      return Optional.of(defaultValue);
    }
    throw LoadException.envVarNotSet(varName, key);
  }

  private static Optional<Object> resolveFile(
      String relativePath, Path agentDir, List<Path> allowedFileRoots, String key) {
    Path requested = Path.of(relativePath);
    Path full = requested.isAbsolute() ? requested : agentDir.resolve(requested);

    Path canonical = canonicalize(full);

    List<Path> roots = new ArrayList<>(allowedFileRoots.size() + 1);
    roots.add(canonicalize(agentDir));
    for (Path root : allowedFileRoots) {
      roots.add(canonicalize(root));
    }

    boolean allowed = roots.stream().anyMatch(canonical::startsWith);
    if (!allowed) {
      throw LoadException.fileReference(
          canonical.toString(),
          "File reference '"
              + relativePath
              + "' for key '"
              + key
              + "' resolves outside allowed roots");
    }

    String content;
    try {
      content = Files.readString(canonical, StandardCharsets.UTF_8);
    } catch (IOException e) {
      throw LoadException.fileReference(canonical.toString(), e.toString());
    }

    String extension = extensionOf(canonical);
    return switch (extension) {
      case "json" -> Optional.of(parseStructured(content, canonical, "Invalid JSON"));
      case "yaml", "yml" -> Optional.of(parseStructured(content, canonical, "Invalid YAML"));
      default -> Optional.of(content);
    };
  }

  /**
   * Parse JSON or YAML content.
   *
   * <p>YAML is a superset of JSON, so SnakeYAML covers both; only the failure wording differs, which
   * keeps diagnostics aligned with the other runtimes.
   */
  private static Object parseStructured(String content, Path path, String errorPrefix) {
    Object parsed;
    try {
      parsed = Frontmatter.newYaml().load(content);
    } catch (YAMLException e) {
      throw LoadException.fileReference(path.toString(), errorPrefix + ": " + e.getMessage());
    }
    if (parsed instanceof Map<?, ?> map) {
      return Frontmatter.stringKeyed(map);
    }
    if (parsed instanceof List<?> list) {
      return Frontmatter.stringKeyedList(list);
    }
    return parsed;
  }

  private static Path canonicalize(Path path) {
    try {
      return path.toRealPath();
    } catch (IOException e) {
      throw LoadException.fileReference(path.toString(), e.toString());
    }
  }

  private static String extensionOf(Path path) {
    String name = path.getFileName().toString();
    int dot = name.lastIndexOf('.');
    return dot < 0 ? "" : name.substring(dot + 1).toLowerCase(Locale.ROOT);
  }
}

File: runtime/swift/prompty/Sources/Prompty/References.swift
import Foundation
import Yams

/// Resolves `${protocol:value}` references in loaded frontmatter.
///
/// Two protocols are supported:
/// - `${env:VAR}` and `${env:VAR:default}` read process environment variables.
/// - `${file:relative/path}` inlines a sibling file. `.json`, `.yaml`, and
///   `.yml` are parsed into structured values; anything else is inlined as text.
///
/// Only whole-value references are resolved — a reference must be the entire
/// string. Unknown protocols are left untouched.
public enum References {

  /// Recursively resolve every reference in a value tree, in place.
  public static func resolve(
    _ value: inout Any,
    agentDirectory: URL,
    allowedRoots: [URL]
  ) throws {
    if var dict = value as? [String: Any] {
      for key in dict.keys {
        if let string = dict[key] as? String {
          if let resolved = try resolveString(
            string, key: key, agentDirectory: agentDirectory, allowedRoots: allowedRoots)
          {
            dict[key] = resolved
          }
        } else if var nested = dict[key] {
          try resolve(&nested, agentDirectory: agentDirectory, allowedRoots: allowedRoots)
          dict[key] = nested
        }
      }
      value = dict
      return
    }

    if var array = value as? [Any] {
      for index in array.indices {
        if let string = array[index] as? String {
          if let resolved = try resolveString(
            string, key: "[\(index)]", agentDirectory: agentDirectory, allowedRoots: allowedRoots)
          {
            array[index] = resolved
          }
        } else {
          var element = array[index]
          try resolve(&element, agentDirectory: agentDirectory, allowedRoots: allowedRoots)
          array[index] = element
        }
      }
      value = array
    }
  }

  /// Resolve a single string. Returns `nil` when the string is not a reference.
  public static func resolveString(
    _ string: String,
    key: String,
    agentDirectory: URL,
    allowedRoots: [URL]
  ) throws -> Any? {
    guard string.hasPrefix("${"), string.hasSuffix("}") else { return nil }

    let inner = String(string.dropFirst(2).dropLast())
    guard let colon = inner.firstIndex(of: ":") else { return nil }

    let proto = inner[inner.startIndex..<colon].lowercased()
    let argument = String(inner[inner.index(after: colon)...])

    switch proto {
    case "env":
      return try resolveEnv(argument, key: key)
    case "file":
      return try resolveFile(
        argument, key: key, agentDirectory: agentDirectory, allowedRoots: allowedRoots)
    default:
      // Unknown protocol — leave the literal in place.
      return nil
    }
  }

  private static func resolveEnv(_ argument: String, key: String) throws -> Any {
    let separator = argument.firstIndex(of: ":")
    let name = separator.map { String(argument[argument.startIndex..<$0]) } ?? argument
    let fallback = separator.map { String(argument[argument.index(after: $0)...]) }

    // An explicitly empty variable is a value, not an absence — deliberately
    // clearing a key must not silently fall back to a default.
    if let value = ProcessInfo.processInfo.environment[name] {
      return value
    }
    if let fallback {
      return fallback
    }
    throw LoadError.envVarNotSet(varName: name, key: key)
  }

  private static func resolveFile(
    _ relativePath: String,
    key: String,
    agentDirectory: URL,
    allowedRoots: [URL]
  ) throws -> Any {
    let requested = URL(fileURLWithPath: relativePath, relativeTo: agentDirectory)
    let full = requested.standardizedFileURL.resolvingSymlinksInPath()

    guard FileManager.default.fileExists(atPath: full.path) else {
      throw LoadError.fileReference(path: full.path, detail: "No such file")
    }

    // File references are a host-controlled capability: by default they may not
    // escape the prompt's own directory tree.
    var roots = [agentDirectory.standardizedFileURL.resolvingSymlinksInPath()]
    roots.append(contentsOf: allowedRoots.map { $0.standardizedFileURL.resolvingSymlinksInPath() })

    let isPermitted = roots.contains { root in
      full.path == root.path || full.path.hasPrefix(root.path + "/")
        || full.path.hasPrefix(root.path + "\\")
    }
    guard isPermitted else {
      throw LoadError.fileReference(
        path: full.path,
        detail: "File reference '\(relativePath)' for key '\(key)' resolves outside allowed roots"
      )
    }

    let contents: String
    do {
      contents = try String(contentsOf: full, encoding: .utf8)
    } catch {
      throw LoadError.fileReference(path: full.path, detail: String(describing: error))
    }

    switch full.pathExtension.lowercased() {
    case "json":
      guard let parsed = JSONSupport.parse(json: contents) else {
        throw LoadError.fileReference(path: full.path, detail: "Invalid JSON")
      }
      return parsed
    case "yaml", "yml":
      do {
        guard let parsed = JSONSupport.normalize(try Yams.load(yaml: contents)) else {
          throw LoadError.fileReference(path: full.path, detail: "Invalid YAML")
        }
        return parsed
      } catch let error as LoadError {
        throw error
      } catch {
        throw LoadError.fileReference(
          path: full.path, detail: "Invalid YAML: \(String(describing: error))")
      }
    default:
      return contents
    }
  }
}
